contrib/endpoints/src/api_manager/api_manager_test.cc

```diff
@@ -249,6 +350,47 @@ TEST_F(ApiManagerTest, InitializedByConfigManager) {
   EXPECT_EQ("2017-05-01r0", service->service().id());
 }
 
+TEST_F(ApiManagerTest, ManagedRolloutStrategy) {
+  std::unique_ptr<MockTimerApiManagerEnvironment> env(
+      new ::testing::NiceMock<MockTimerApiManagerEnvironment>());
+
+  EXPECT_CALL(*env.get(), DoRunHTTPRequest(_))
+      .WillOnce(Invoke([this](HTTPRequest *req) {
+        ASSERT_EQ(
+            "https://servicemanagement.googleapis.com/v1/services/"
+            "bookstore.test.appspot.com/rollouts?filter=status=SUCCESS",
+            req->url());
+        req->OnComplete(Status::OK, {}, kRolloutsResponse1);
+      }))
+      .WillOnce(Invoke([this](HTTPRequest *req) {
+        ASSERT_EQ(
+            "https://servicemanagement.googleapis.com/v1/services/"
+            "bookstore.test.appspot.com/configs/2017-05-01r1",
+            req->url());
+        req->OnComplete(Status::OK, {}, kServiceConfig2);
+      }));
+
+  std::shared_ptr<ApiManagerImpl> api_manager(
+      std::dynamic_pointer_cast<ApiManagerImpl>(MakeApiManager(
+          std::move(env), kServerConfigWithManagedRolloutStrategy)));
+
+  EXPECT_TRUE(api_manager);
+  EXPECT_TRUE(api_manager->Enabled());
+  EXPECT_EQ("bookstore.test.appspot.com", api_manager->service_name());
+  EXPECT_EQ("2017-05-01r0", api_manager->service("2017-05-01r0").id());
+
+  api_manager->Init();
+
+  EXPECT_TRUE(api_manager->Enabled());
+  EXPECT_EQ("2017-05-01r0", api_manager->service("2017-05-01r0").id());
+
+  auto service = api_manager->SelectService();
```

Comment:
In this test, ApiManager was initialized in the constructor and Init() function. 

When ApiManager creates an instance of ConfigManager in Init(), the constructor of ConfigManager invokes the callback.

https://github.com/istio/proxy/pull/375/files/3529d1ef9d24ececcb583f27ebc52acbf9a70bc7#diff-43539418cb3efc25680906e605c5c19eR189